### PR TITLE
[MISC] Fix version juju==3.0.4 and formatting fixes

### PR DESCRIPTION
**Pull request by deusebio**

### CODE REVIEW

**Review by arturo-seijas — APPROVED** [review]
LGTM

File: src/metadata.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def get(path: Path) -> types_.Metadata:
         raise InputError(f"Invalid value for name key: {name}, expected a string value")
 
     docs = metadata.get(METADATA_DOCS_KEY)
-    if not isinstance(docs, str | None) or (METADATA_DOCS_KEY in metadata and docs is None):
+    if docs and not isinstance(docs, str) or (METADATA_DOCS_KEY in metadata and docs is None):
```
Comment (jdkandersson) [inline src/metadata.py:59]: As noted on the other PR, This change leads to to rejecting empty `docs` values which is fine. Should the error message also be updated?

**Review by jdkandersson — COMMENTED** [review]
I'll change the target to be a separate branch so that the e2e tests can run

**Comment by jdkandersson** [comment]
Also looks like there is a bug in postgres, I'll see if I have time to look into that

**Comment by jdkandersson** [comment]
> Also looks like there is a bug in postgres, I'll see if I have time to look into that

I have looked into it a bit and it seems there is some problem with disocurse-k8s which fails when the database relation is joined. Haven't progressed beyond that

**Comment by yanksyoon** [comment]
The Unit In Error is currently a bug on juju, please refer to the following for more information!
PR: https://github.com/juju/python-libjuju/pull/807
CI Test: https://github.com/jnsgruk/kubernetes-dashboard-operator/actions/runs/4321658165/jobs/7543594265
Discussion on public mattermost: https://chat.charmhub.io/charmhub/pl/zkix8s5jjfngpk5yp4t1f55u4e

**Comment by jdkandersson** [comment]
As discussed, the remedy seems to be either (1) wait for the fix or (2) go back to pre juju 3

**Review by yanksyoon — APPROVED** [review]
LGTM, may I suggest that we fix the juju version to versions less than 3, for example 2.9.38/2.9.11?
Due to issues with juju versions the integration tests are failing and the workaround is to downgrade the juju version for now. Here are some of the examples of the repositories that are not using juju versions >3
[opensearch operator](https://github.com/canonical/opensearch-operator/blob/a83dab094743cca4ce27712aff7e7bc035849c21/tox.ini#L107), [kafka operator](https://github.com/canonical/kafka-operator/blob/0352752fc3ef6a28f1e54a93db849f3429912ace/pyproject.toml#L68)

**Comment by yanksyoon** [comment]
@deusebio Thank *you*! 🚀

**Comment by deusebio** [comment]
@yanksyoon I'm trying to understand how to fix the agent version of juju to 2.9.38. In the kafka-repo this is done [here](https://github.com/canonical/kafka-operator/blob/0352752fc3ef6a28f1e54a93db849f3429912ace/.github/workflows/ci.yaml#L72), but I see that in this repo we inherit other actions from [here](https://github.com/canonical/operator-workflows/blob/main/.github/workflows/integration_test.yaml) but I'm not sure where I can provide this flag there.

**Comment by yanksyoon** [comment]
On my local environment, I was able to get the integration tests run up to discourse API key creation failure
```
 AssertionError: discourse main API key creation failed, stderr='/tmp/juju-exec10631795
```

There seems to be some bugs due to new discourse release (promoted to stable on 28th Feb).
1. [THROTTLE_LEVELS](https://github.com/canonical/discourse-k8s-operator/blob/main/src/charm.py#L37) mapping to non-str causes pebble error (env should be str:str mapping) -> this is an issue with discourse charm
2. Since it's a sidecar charm the fixture `discourse_main_api_key` doesn't work anymore. The API key creation script is supposed to be run on `discourse` container, not the charm container.

I've tried to run the api creation script 
```
cd /srv/discourse/app && ./bin/bundle exec rake api_key:create_master['main API key for testing'] RAILS_ENV=production
```
with correct environment variables, but i'm getting error with the script trying to connect to postgres
```
ActiveRecord::ConnectionNotEstablished: could not connect to server: No such file or directory
Is the server running locally and accepting
connections on Unix domain socket "/var/run/postgresql/.s.PGSQL.5432"?
```
but my environment settings should be okay(pasted below). I'll do a bit more looking on how this script works and why it's trying to connect through the unix domain socket.
```
HOSTNAME=discourse-k8s-0
CONTAINER_APP_VERSION=v2.7.10
CONTAINER_APP_UID=200
CONTAINER_APP_GROUP=discourse
CONTAINER_APP_GID=200
DEBIAN_FRONTEND=noninteractive
PLUGINS_DIR=/srv/discourse/app/plugins
JUJU_CONTAINER_NAME=discourse
PEBBLE_SOCKET=/charm/container/pebble.socket
POSTGRESQL_K8S_OPERATOR_PORT_30666_TCP_PORT=30666
MODELOPERATOR_SERVICE_PORT=17071
MODELOPERATOR_PORT_17071_TCP_ADDR=10.152.183.245
KUBERNETES_SERVICE_HOST=10.152.183.1
KUBERNETES_SERVICE_PORT=443
DISCOURSE_K8S_PORT_65535_TCP_PORT=65535
POSTGRESQL_K8S_OPERATOR_PORT=tcp://10.152.183.236:30666
POSTGRESQL_K8S_OPERATOR_PORT_30666_TCP_PROTO=tcp
REDIS_K8S_OPERATOR_SERVICE_HOST=10.152.183.218
REDIS_K8S_OPERATOR_PORT_30666_TCP_PORT=30666
DISCOURSE_K8S_PORT_65535_TCP=tcp://10.152.183.157:65535
REDIS_K8S_OPERATOR_PORT=tcp://10.152.183.218:30666
MODELOPERATOR_PORT_17071_TCP=tcp://10.152.183.245:17071
KUBERNETES_PORT_443_TCP_PORT=443
DISCOURSE_K8S_PORT_65535_TCP_PROTO=tcp
REDIS_K8S_OPERATOR_SERVICE_PORT=30666
KUBERNETES_PORT_443_TCP_PROTO=tcp
DISCOURSE_K8S_SERVICE_PORT_PLACEHOLDER=65535
DISCOURSE_K8S_PORT=tcp://10.152.183.157:65535
MODELOPERATOR_SERVICE_HOST=10.152.183.245
KUBERNETES_PORT_443_TCP=tcp://10.152.183.1:443
POSTGRESQL_K8S_OPERATOR_PORT_30666_TCP_ADDR=10.152.183.236
MODELOPERATOR_PORT=tcp://10.152.183.245:17071
MODELOPERATOR_PORT_17071_TCP_PROTO=tcp
KUBERNETES_SERVICE_PORT_HTTPS=443
KUBERNETES_PORT=tcp://10.152.183.1:443
REDIS_K8S_OPERATOR_PORT_30666_TCP=tcp://10.152.183.218:30666
POSTGRESQL_K8S_OPERATOR_SERVICE_PORT=30666
POSTGRESQL_K8S_OPERATOR_PORT_30666_TCP=tcp://10.152.183.236:30666
REDIS_K8S_OPERATOR_PORT_30666_TCP_PROTO=tcp
REDIS_K8S_OPERATOR_PORT_30666_TCP_ADDR=10.152.183.218
DISCOURSE_K8S_SERVICE_PORT=65535
DISCOURSE_K8S_PORT_65535_TCP_ADDR=10.152.183.157
KUBERNETES_PORT_443_TCP_ADDR=10.152.183.1
DISCOURSE_K8S_SERVICE_HOST=10.152.183.157
POSTGRESQL_K8S_OPERATOR_SERVICE_HOST=10.152.183.236
MODELOPERATOR_PORT_17071_TCP_PORT=17071
HOME=/root
CONTAINER_APP_NAME=discourse
DISCOURSE_REDIS_PORT=6379
DISCOURSE_REDIS_HOST=10.1.158.114
DISCOURSE_DEVELOPER_EMAILS=
DISCOURSE_SMTP_USER_NAME=
DISCOURSE_DB_NAME=discourse
DISCOURSE_HOSTNAME=discourse
DISCOURSE_SMTP_PASSWORD=
DISCOURSE_SERVE_STATIC_ASSETS=true
DISCOURSE_DB_HOST=10.152.183.60
DISCOURSE_SMTP_DOMAIN=
DISCOURSE_MAX_REQS_PER_IP_MODE=warn+block
DISCOURSE_ENABLE_CORS=true
DISCOURSE_MAX_REQS_PER_IP_PER_MINUTE=1000
DISCOURSE_CORS_ORIGIN=*
DISCOURSE_SMTP_ADDRESS=
DISCOURSE_MAX_USER_API_REQS_PER_MINUTE=400
DISCOURSE_SMTP_OPENSSL_VERIFY_MODE=none
DISCOURSE_DB_USERNAME=discourse-k8s
DISCOURSE_SMTP_PORT=587
DISCOURSE_REFRESH_MAXMIND_DB_DURING_PRECOMPILE_DAYS=0
DISCOURSE_MAX_REQS_PER_IP_PER_10_SECONDS=100
RAILS_ENV=production
DISCOURSE_SMTP_AUTHENTICATION=none
DISCOURSE_DB_PASSWORD=ZKRFw2WrRj5YFWkwk2FfwNKKHpGXL7FVcmjfgZqh
GEM_HOME=/srv/discourse/.gem
DISCOURSE_MAX_ASSET_REQS_PER_IP_PER_10_SECONDS=400
CONTAINER_APP_USERNAME=discourse
CONTAINER_APP_ROOT=/srv/discourse
DISCOURSE_MAX_REQS_RATE_LIMIT_ON_PRIVATE=false
```

**Comment by deusebio** [comment]
Thanks @yanksyoon for this update and for looking into this. I'm wondering, is there any way to force to use a different charm revision/version? Just because I believe here we are actually starting to go well beyond the original scope of this PR into a more severe refactoring. 

I'm just wondering whether there could be any easy way (or workarounds) to get things working (although not with most recent versions of charms/juju)

Otherwise this starts to be rather blocking for everything else we want to start to work on in the next weeks to implement sync processes between Discourse and Github (tagging @jdkandersson for alignment here)

**Comment by jdkandersson** [comment]
I think we can close this now since it has been incorporated in another PR

**Pull request closed without merging**